src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java

```diff
@@ -0,0 +1,83 @@
+package com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeSmell;
+
+/*
+ * This file is part of the Php Inspections (EA Extended) package.
+ *
+ * (c) Funivan <alotofall@gmail.com>
+ * (c) Vladimir Reznichenko <kalessil@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+import com.intellij.codeInspection.ProblemHighlightType;
+import com.intellij.codeInspection.ProblemsHolder;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiElementVisitor;
+import com.intellij.psi.PsiRecursiveElementWalkingVisitor;
+import com.jetbrains.php.lang.psi.elements.Function;
+import com.jetbrains.php.lang.psi.elements.Method;
+import com.jetbrains.php.lang.psi.elements.PhpClass;
+import com.jetbrains.php.lang.psi.elements.PhpReturn;
+import com.jetbrains.php.lang.psi.visitors.PhpElementVisitor;
+import com.kalessil.phpStorm.phpInspectionsEA.openApi.BasePhpInspection;
+import com.kalessil.phpStorm.phpInspectionsEA.utils.ExpressionSemanticUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class MultipleReturnPointsInspector extends BasePhpInspection {
+    private static final String message = "Method with multiple return points.";
+
+
+    @NotNull
+    public String getShortName() {
+        return "MultipleReturnPointsInspection";
+    }
+
+    @NotNull
+    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
+        return new PhpElementVisitor() {
+
+            public void visitPhpMethod(Method method) {
+                final PhpClass containingClass = method.getContainingClass();
+                final PsiElement nameIdentifier = method.getNameIdentifier();
+                if (containingClass != null && nameIdentifier != null && !containingClass.isTrait()) {
+
+                    final ReturnPointCountVisitor countVisitor = new ReturnPointCountVisitor(method);
```

Comment:
As I mentioned, that a recursive visitor must be avoided, I should provide an alternative. In the context of this task, I'd recommend using a control flow. It can be easily obtained calling `method.getControlFlow()`. A building of a control flow is a kinda expensive task, but once built it's cached and can be reused by several inspections. So most probably some other inspection is already built it for you and you'll get it for free. (Disclaimer: never trust such sentences and always re-check using a profiler.)

Here's the sketch how the code can look like.

    final PhpControlFlow controlFlow = method.getControlFlow();
    final PhpExitPointInstruction exitPoint = controlFlow.getExitPoint();
    int count = 0;
    for (final PhpInstruction instruction : exitPoint.getPredecessors()) {
      if (instruction instanceof PhpReturnInstruction) {
        count++;
      }
      if (count > 1) {
        holder.registerProblem(...);
        return;
      }
    }

Most probably this code doesn't cover some corner cases like `finally`. So you should check how does it work and adjust accordingly. The `PhpControlFlowUtil` class can be helpful. (Note: it's not a part of the open API thus a subject to change.)